common/safety-push-metrics/src/pusher.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::counters::Metrics;
+use libra_logger::*;
+use std::{env, sync::Arc, thread, thread::JoinHandle, time::Duration};
+
+const DEFAULT_PUSH_FREQUENCY_SECS: u64 = 15;
+
+pub struct MetricsPusher {
+    metrics: Arc<dyn Metrics + Send + Sync>,
+}
+
+impl MetricsPusher {
+    pub fn new(metrics: Arc<dyn Metrics + Send + Sync>) -> Self {
+        Self { metrics }
+    }
+
+    pub fn start(&'static self) -> JoinHandle<()> {
+        thread::spawn(move || {
+            // eg value for PUSH_METRICS_ENDPOINT: "http://pushgatewar.server.com:9091/metrics/job/safety_rules"
+            let push_metrics_endpoint = match env::var("PUSH_METRICS_ENDPOINT") {
+                Ok(s) => s,
+                Err(_) => {
+                    info!("PUSH_METRICS_ENDPOINT env var is not set. Skipping sending metrics.");
+                    return;
+                }
+            };
+            let push_metrics_frequency_secs = match env::var("PUSH_METRICS_FREQUENCY_SECS") {
+                Ok(s) => match s.parse::<u64>() {
+                    Ok(i) => i,
+                    Err(_) => {
+                        error!("Invalid value for PUSH_METRICS_FREQUENCY_SECS: {}", s);
+                        return;
+                    }
+                },
+                Err(_) => DEFAULT_PUSH_FREQUENCY_SECS,
+            };
+            info!(
+                "Starting push metrics loop. Sending metrics to {} with a frequency of {} seconds",
+                push_metrics_endpoint, push_metrics_frequency_secs
+            );
+            loop {
+                let mut data = "".to_string();
+                for metric in self.metrics.get_metrics() {
+                    let metric_data = metric.get_push_metric_format();
+                    data.push_str(&metric_data);
+                }
+                debug!("Metrics sent to server:\n{}", data);
+                let response = ureq::post(&push_metrics_endpoint)
+                    .timeout_connect(10_000)
+                    .send_string(&data);
+
+                if response.error() {
+                    warn!("Failed to push metrics to {}", push_metrics_endpoint);
```

Comment:
could we print error?

Reply:
Done

Reply:
I should have said print /the/ error warn / error as the logging level is fine :P